core/lib/zksync_core/src/consistency_checker/mod.rs

```diff
@@ -1,204 +1,318 @@
-use std::time::Duration;
+use std::{fmt, time::Duration};
 
+use anyhow::Context as _;
+use async_trait::async_trait;
+use tokio::sync::watch;
 use zksync_contracts::PRE_BOOJUM_COMMIT_FUNCTION;
 use zksync_dal::ConnectionPool;
 use zksync_types::{
-    web3::{error, ethabi, transports::Http, types::TransactionId, Web3},
-    L1BatchNumber,
+    web3::{self, ethabi, transports::Http, types::TransactionId, Web3},
+    L1BatchNumber, H256, U64,
 };
 
-use crate::metrics::{CheckerComponent, EN_METRICS};
+use crate::{
+    metrics::{CheckerComponent, EN_METRICS},
+    utils::wait_for_l1_batch_with_metadata,
+};
+
+#[cfg(test)]
+mod tests;
+
+#[derive(Debug, thiserror::Error)]
+enum CheckError {
+    #[error("Web3 error communicating with L1")]
+    Web3(#[from] web3::Error),
+    #[error("Internal error")]
+    Internal(#[from] anyhow::Error),
+}
+
+impl From<zksync_dal::SqlxError> for CheckError {
+    fn from(err: zksync_dal::SqlxError) -> Self {
+        Self::Internal(err.into())
+    }
+}
+
+#[async_trait]
+trait L1Client: fmt::Debug + Send + Sync {
```

Comment:
Do we need a separate trait for that? Wouldn't `eth_client::EthInterface` suffice here?

Reply:
`EthInterface` cannot be made into a trait object because of the parametric `call_contract_function` method. IMO, it'd make sense to refactor this method so that it's not parametric, and use an extension trait for the parametric version (optionally, add a builder pattern as well to manage argument count). Would you prefer me doing this in this PR, or separately? As a possible alternative (IMO, a suboptimal one), `ConsistencyChecker` itself may incorporate a `C: EthInterface` type param.

Reply:
As discussed, would love to see object-safe implementation of `EthInterface`.